### PR TITLE
[python] Implement `IntIndexer` as class that wraps around `clib.IntIndexer`

### PR DESCRIPTION
**Pull request by jp-dark**
**Issue and/or context:** #2187

**Changes:**
* Replace the `tiledbsoma_build_index` function with a new `IntIndex` class that wraps `clib.IntIndexer`.
* Move type checking from PyBind interface in the `clib.IntIndexer` class to the python layer using `isinstance`.

**Notes for Reviewer:**

Required a bump in somacore to relax the requirements for an `IndexLike` class.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -15,7 +15,7 @@ repos:
     - id: mypy
       additional_dependencies:
         - "pandas-stubs==1.5.3.230214"
-        - "somacore==1.0.9"
+        - "somacore==1.0.10"
```
Comment (johnkerl) [inline .pre-commit-config.yaml:18]: Also please `.pre-commit-config.yaml`

File: apis/python/src/tiledbsoma/_indexer.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+"""
+This file is separate from _util.py, due to a circular-import issue with
+SOMATileDBContext which would otherwise ensue.
+"""
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/_indexer.py:4]: Can remove this comment now.

File: apis/python/src/tiledbsoma/_indexer.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        self,
+        data: Union[  # type: ignore[type-arg]
+            npt.NDArray[np.int64],
+            pa.Array,
+            pa.IntegerArray,
+            pd.Series,
+            pd.arrays.IntegerArray,
+            pa.ChunkedArray,
+            list[int],
+        ],
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/_indexer.py:38]: Maybe we can have a const `IndexerDataType = Union[...]`.

File: apis/python/src/tiledbsoma/reindexer.cc [hunk 4]
```diff
@@ -148,15 +148,17 @@ void load_reindexer(py::module& m) {
         // looked up values into previously allocated array (works for the
         // cases in which python and R pre-allocate the array)
         .def(
-            "get_indexer",
+            "get_indexer_general",
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/reindexer.cc:151]: Are we able to just do `.def("get_indexer_general", &IntIndexer::get_indexer_general)` now?

File: apis/python/src/tiledbsoma/reindexer.cc [hunk 5]
```diff
@@ -148,15 +148,17 @@ void load_reindexer(py::module& m) {
...
             [](IntIndexer& indexer, py::array_t<int64_t> lookups) {
                 return get_indexer_general(indexer, lookups);
             })
         // If the input is not arrow (does not have _export_to_c attribute),
         // it will be handled using a general input method.
-        .def("get_indexer", [](IntIndexer& indexer, py::object py_arrow_array) {
-            return get_indexer_py_arrow(indexer, py_arrow_array);
-        });
+        .def(
+            "get_indexer_pyarrow",
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/reindexer.cc:158]: Same comment as above.

File: apis/python/tests/test_reindexer_api.py [hunk 6]
```diff
@@ -1,22 +1,16 @@
...
-    indexer = tiledbsoma_build_index(keys)
-    result = indexer.get_indexer(ids)
-    assert np.equal(result.all(), expected.all())
+from tiledbsoma import IntIndexer, SOMATileDBContext
 
 
-def test_reindexer_api_context():
-    context = SOMATileDBContext()
+@pytest.mark.parametrize("context", [None, SOMATileDBContext()])
+def test_reindexer_api(context: Optional[SOMATileDBContext]) -> None:
```
Comment (nguyenv) [inline apis/python/tests/test_reindexer_api.py:10]: We don't usually have a return typing hint for the unit tests.

**Review by nguyenv — APPROVED** [review]
Thank you so much! Looks overall good to me. Just some minor comments.

**Comment by jp-dark** [comment]
[sc-43678]

**Pull request merged**